content/posts/2019-07-21-oop-and-fp.md

```diff
@@ -0,0 +1,174 @@
+Nothing about classes, or types. In fact, right after the defintion of OOP, Kay also says this:
+
+> I'm not against types, but I don't know of any type systems that aren't a complete pain, so I still like dynamic typing.
+
+It's fascinating to see people advocating for strong typing in OOP contexts, while the person who coined the term prefers dynamic typing. Enough with this tangent, back to OOP. Let's dissect each piece of the definition.
+
+## Messaging
+
+Messaging is a core concept in OOP. In class-based languages, they usually come in forms on method calls, but it doesn't mean that every method call is a message, as pointed out in the [MOOM course](https://avdi.codes/courses/moom/).
+
+In Go, for instance, you can think of messages as [_channels_](https://gobyexample.com/channels), which is a mechanism to achieve communication between goroutines. In Elixir, think of [GenServer](https://elixir-lang.org/getting-started/mix-otp/genserver.html). You start a process (something like very light-weight thread of execution) that behaves like it's own "server in the network" and receives messages from other processes and responds back to those processes with other messages (we will see an example of this later on).
+
+## Local Retention and Protection and Hiding of State-Process
+
+Here we are talking about encapsulation of state. Be it on instances in class-based languages, or processes in concurrent languages. The idea is that the "sender" of the message (the caller of the method or the process sending the message), has no idea of what data is hold in the internals of the message handler. It couldn't care less. All it knows is that by sending a message in this format, it can receive a response (or maybe not) back.
+
+## Extreme late-binding
+
+Let's see what [Wikipedia](https://en.wikipedia.org/wiki/Late_binding) has o say about late-binding:
+
+> Late binding, dynamic binding, or dynamic linkage is a computer programming mechanism in which the method being called upon an object or the function being called with arguments is looked up by name at runtime.
+
+In other words, late-binding is waiting until the last possible moment (at runtime) to decide how a _message_ is going to be handled. It's contrary to the idea of static binding (or early binding), usually (but not exclusive to) from statically-typed, compiled languages where a compiler binds at compilation time how the messages sent in the system will be executed and which objects/functions are going to handle it. It gives very little to figure out at runtime.
+
+Many patterns emerged in order to achieve this in statically-typed languages, like [Inversion of Control](https://en.wikipedia.org/wiki/Inversion_of_control). Instead of statically-binding to a concrete implementation, you bind the dependency to an interface and let *something* for the runtime to figure out (via configuration, factory pattern, or what ever) which instance that implements the interface is going to handle that message.
```

Comment:
I see interfaces as another way to implement late binding. What do you think?

In this paragraph, you focus on ways to use interfaces (the patterns) but you could follow the next paragraphs and get a nice flat view of interfaces x duck typing x generics trying to solve late binding with their own goals and means. 

Maybe it is also important to note that they could co-exist. C++, Java and [soon Go](https://blog.golang.org/why-generics) will have both generics and interfaces.